in-progress/8025-publishing-tx-objects/design.md

```diff
@@ -48,27 +69,159 @@ To the DA oracle contract, the proposer will publish the following data:
   - Tx2
   - ...
 
+> Most of the global variables could be directly populated by L1, but providing it makes it a lot clearer what the reason of failure is if such a one is encountered. Also the `inHash` can be populated by L1.
 
-To the rollup contract, the proposer will publish:
+#### DA Oracle Changes
 
-- CallData
-  - numTxs
-  - txsHash (a commitment to first nullifiers of all transactions in the (eventual) block)
-  - inHash
-  - GlobalVariables
-    - slotNumber
-    - timestamp
-    - coinbase
-    - feeRecipient
+When the data is to be submitted as part of the same transaction as the block proposal to the L1 contract, we have two variations possible:
 
-#### DA Oracle Changes
+1. Remove the data availability oracle in its whole, and simply use the versioned KZG hashes directly
+2. Save the versioned KZG hashes in the data availability oracle
 
-Keeps track of the versioned KZG hashes that were present in a transaction.
+Solution 1 is significantly cheaper as it don't update storage. Furthermore, by requiring that the versioned KZG hashes are part of the header provided, we can show the exact KZG hashes using just the transaction and the commitment to the block hash, see `BlockLog.blockHash`.
 
 #### Rollup Contract Changes
+> There is probably some 1-off from the `count` `block_number` like values.
+
+```python
+struct State:
+  pending_proposal_count: uint64;
+  pending_slot: uint64;
+  proven_block_count: uint64;
+  proven_slot: uint64;
+
+struct ProposalLog:
+  proposal_hash: bytes32;
+  slot: uint256;
+  archive: bytes32;
+
+proposals: public(HashMap[uint256, ProposalLog])
+state: public(State);
+
+# The old `process`
+def propose(proposal, proposer_sig, attestations):
+  '''
+  Notice that no signatures are checked, all of that is optimistic 
+  '''
+  assert hash([h for h in tx.blob_versioned_hashes]) == proposal.kzgHashes # Replaces DA oracle
+  assert proposal.gas_settings.is_sane() # To be designed
+  assert proposal.slot > state.pending_slot and proposal.slot == get_current_slot()
+  assert proposal.timestamp == get_time_at_slot(proposal.slot) and proposal.timestamp <= block.timestamp
+  assert proposal.block_number == state.pending_proposal_count
+  assert proposal.inHash == INBOX.consume(proposal.block_number)
+
+  self.proposals[proposal.block_number] = ProposalLog(sha256(proposal,  proposer_sig, attestations), proposal.slot)
+  self.state.pending_proposal_count += 1
+  self.state.pending_slot = proposal.slot
+
+def challenge_proposal(proposal, proposer_sig, attestations):
+  '''
+  Implement challenges for:
+  - Bad proposer
+  - Bad attestation
+  '''
+  pass
+
+
+def submit_proof(proof, archive):
+  '''
+  Notice that this only supports sequential proofs as we always feed `self.state.proven_archive` as start state.
+  '''
+  proposal_hashes = []
+  block_number_start = self.state.proven_block_count
+  end_slot = get_last_slot_in_epoch(get_epoch_from_slot(self.proposals[block_number_start].slot))
```

Comment:
I think this should be 
```python
get_last_slot_in_epoch(get_epoch_from_slot(self.proposals[block_number_start].slot + 1))
```
assuming we want the last slot in the epoch after the most recently proven epoch.

Reply:
The index/count is messing with you. Will try to fix them for less confusion